server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -617,6 +625,21 @@ public Object call() throws IOException
     // NB: The rows are still in memory until they're done persisting, but we only count rows in active indexes.
     rowsCurrentlyInMemory.addAndGet(-numPersistedRows);
     bytesCurrentlyInMemory.addAndGet(-bytesPersisted);
+
+    log.info("Persisted rows[%,d] and bytes[%,d]", numPersistedRows, bytesPersisted);
+
+    // bytesCurrentlyInMemory can change while persisting due to concurrent ingestion.
+    // Hence, we use bytesInMemoryBeforePersist to determine the change of this persist
+    if (bytesInMemoryBeforePersist - bytesPersisted > maxBytesTuningConfig) {
+      // We are still over maxBytesTuningConfig even after persisting.
+      // This means that we ran out of all available memory to ingest (due to overheads created as part of ingestion)
+      log.makeAlert("Persist no longer free up memory")
+         .addData("dataSource", schema.getDataSource())
+         .addData("numSinks", sinks.size())
+         .addData("numHydrantsAcrossAllSinks", sinks.values().stream().mapToInt(Iterables::size).sum())
+         .emit();
+      throw new RuntimeException("Ingestion run out of available memory as persist can no longer free up any memory.");
+    }
```

Comment:
Maybe this should clarify that the task is using too much heap memory? I also suggest moving `numSinks ` and `numHydrantsAcrossAllSinks` into the error messaging and sharing the error message between the alert and the exception that is thrown after so that it is a bit more visible. I also wonder if we should include the total row count to aid in determining if perhaps overly aggressive heap memory limits caused a large number of persists, or if it is just a large amount of data.

Thinking out loud (and not something we really need to worry about in this PR), I wonder if we should perhaps consider more targeted error messaging. It seems like there are at least 4 ways we end up here:
* too many intervals/sinks
* too many persisted segments in a reasonable number of sinks and rows and columns (too much data)
* too many columns in a reasonable number of segments and sinks and rows
* too many persisted segments in a reasonable number of sinks and columns but a unreasonable/small number of total rows (aggressive settings problem)
 
The trick I guess would be deciding what is "reasonable" to adjust the error messaging accordingly to suggest the source of the problem, though is that actually useful? The cluster operator can infer any of these situations this from whether the `numSinks ` and `numHydrantsAcrossAllSinks` and `totalRows` are large numbers or not, so maybe it isn't worth the extra messaging, since there isn't a distinct or obvious way to mitigate any of these conditions.

Reply:
Done. Added more details to error message. Moved stats into error message. Added totalRow stat.